### PR TITLE
Sam4l flash controller implementation

### PR DESCRIPTION
**Pull request by KBaichoo**

### CODE REVIEW

File: src/platform/storm/lib.rs [hunk 1]
```diff
@@ -26,6 +26,7 @@ pub mod io;
 //mod gpio_dummy;
 //mod spi_dummy;
 //mod i2c_dummy;
+mod flash_dummy;
```
Comment (alevy) [inline src/platform/storm/lib.rs]: comment out

File: src/chips/sam4l/flashcalw.rs [hunk 2]
```diff
@@ -0,0 +1,1106 @@
...
+        self.page.set(page_num);
+        self.current_state.set(FlashState::Unlocking);
+        self.current_command.set(Command::Erase);
+        self.lock_page_region(page_num, false);
+        0 
+    }
+}
+
+//  Assumes the only Hardware Interrupt enabled for the FLASHCALW is the
+//  FRDY (Flash Ready) interrupt.
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: public comments in Rust (e.g. shown in documentation) have three forward slashes (`///`) in front of them.

File: src/chips/sam4l/flashcalw.rs [hunk 3]
```diff
@@ -0,0 +1,1106 @@
+/* chips::sam4l::flashcalw -- Implementation of a flash controller.
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Module comments are `//!`

File: src/chips/sam4l/flashcalw.rs [hunk 4]
```diff
@@ -0,0 +1,1106 @@
...
+ */
+
+use helpers::*;
+use core::{mem, slice};
+use core::cell::Cell;
+use common::take_cell::TakeCell;
+use pm;
+use nvic;
+
+//  These are the registers of the PicoCache -- a cache sole for the flash.
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: what is a cache sole?

File: src/chips/sam4l/flashcalw.rs [hunk 5]
```diff
@@ -0,0 +1,1106 @@
...
+#[allow(dead_code)]
+struct Picocache_Registers {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: CamelCase

File: src/chips/sam4l/flashcalw.rs [hunk 6]
```diff
@@ -0,0 +1,1106 @@
...
+    reserved_1:                             [u8;0x8],
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: this is fine, but in general, unused fields should be named `_fieldname`, so for example `_reserved1`

Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: * a space after the semicolon
 * Decimal array size

```rust
[u8; 8]
```

File: src/chips/sam4l/flashcalw.rs [hunk 7]
```diff
@@ -0,0 +1,1106 @@
...
+    picocache_control:                      usize,
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: the `picocache_` prefixes are not necessary, since these are already fields of the picocache struct

**Comment by alevy** [comment]
Architecturally, I think the picocache registers structure should a field of the `Registers` struct or another field of the Flashcalw struct. In any case the functions that operate on the picocache should be methods of Flashcalw and the address of the picocache should be an offset relative to the flashcalw base address rather than separately hard coded.

File: src/chips/sam4l/flashcalw.rs [hunk 8]
```diff
@@ -0,0 +1,1106 @@
...
+    picocache_status:                       usize,
+    reserved_2:                             [u8;0x10],
+    picocache_maintenance_register_0:       usize,
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: In general, we're moving to explicit `u32` as the underlying register type.

Also, you should use `VolatileCell` instead which will allow you to avoid mutable references in many places.

File: src/chips/sam4l/flashcalw.rs [hunk 9]
```diff
@@ -0,0 +1,1106 @@
...
+}
+
+pub fn pico_enabled() -> bool {
+    let registers : &mut Picocache_Registers = unsafe { 
+        mem::transmute(PICOCACHE_BASE_ADDRS)
+    };
+   volatile_load(&registers.picocache_status) & 0x1 != 0
+}
+
+// Listing of the FLASHCALW register memory map.
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: It's not a listing, it's an actual struct.

File: src/chips/sam4l/flashcalw.rs [hunk 10]
```diff
@@ -0,0 +1,1106 @@
...
+    COMMAND,
+    STATUS,
+    PARAMETER,
+    VERSION,
+    GPFRHI,
+    GPFRLO
+}
+
+/// ERROR codes
+pub enum Error {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Could you document the meaning of each error code?

File: src/chips/sam4l/flashcalw.rs [hunk 11]
```diff
@@ -0,0 +1,1106 @@
...
+    CommandComplete,
+    LockE,
+    ProgE,
+    LockProgE,
+    ECC,
+}
+
+// High level commands to issue to the flash.
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Perhaps useful to document what the purpose of these commands are relative to `FlashCMD`

**Comment by alevy** [comment]
OK, it doesn't merge cleanly, so you need to merge manually

**Comment by KBaichoo** [comment]
Hey @alevy I've address your first rounds of comments and have merged with master. Can you give it a look now?

File: src/chips/sam4l/lib.rs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 #![crate_name = "sam4l"]
 #![crate_type = "rlib"]
-#![feature(asm,core_intrinsics,concat_idents,const_fn)]
+#![feature(lang_items,asm,core_intrinsics,concat_idents,const_fn, clone_from_slice)]
```
Comment (alevy) [inline src/chips/sam4l/lib.rs]: you don't need `lang_items`

File: src/chips/sam4l/flashcalw.rs [hunk 13]
```diff
@@ -0,0 +1,1103 @@
...
+}
+
+/// This is simply std::cmp::min from std
+#[inline]
+fn min<T: Ord>(v1: T, v2: T) -> T {
+    if v1 <= v2 { v1 } else { v2 }    
+}
+
+// Should never ever have to wait (because we're using interrupts!)
+pub fn default_wait_until_ready(flash : &FLASHCALW) {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Hrm... so why ever call this?

Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Hrm... so why ever call this?

File: src/chips/sam4l/flashcalw.rs [hunk 14]
```diff
@@ -0,0 +1,1103 @@
...
+const GP_ALL_FUSES_ONE : u64 = !0 as u64;
+
+// Macros for getting the i-th bit.
+macro_rules! get_bit {
+    ($w:expr) => (0x1u32 << $w);
+}
+
+/// This is simply std::cmp::min from std
+#[inline]
+fn min<T: Ord>(v1: T, v2: T) -> T {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: just use `core::cmp::min`. most standard library functions are really just re-exports of functions in libcore

File: src/chips/sam4l/flashcalw.rs [hunk 15]
```diff
@@ -0,0 +1,1103 @@
...
+const FLASH_FREQ_PS1_FWS_0_MAX_FREQ : u32 = 8000000;
+
+
+#[cfg(not(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE))]
+const FLASH_FREQ_PS2_FWS_0_MAX_FREQ : u32 = 24000000;
+
+const GP_ALL_FUSES_ONE : u64 = !0 as u64;
+
+// Macros for getting the i-th bit.
+macro_rules! get_bit {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: I would call this `bit!`. `get_bit` implies to me that you're getting something from the passed in argument.

File: src/chips/sam4l/flashcalw.rs [hunk 16]
```diff
@@ -0,0 +1,1103 @@
...
+
+/// Trait for a client of the flash controller.
+pub trait Client {
+    //  Called upon a completed call
+    fn command_complete(&self, err: Error);     
+}
+    
+impl FLASHCALW {
+
+    pub fn mark_ready(&self) {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Why expose this? It also seems to only be called in a single place `handle_interrupt` and is just one line, so maybe just put the line there?

File: src/chips/sam4l/flashcalw.rs [hunk 17]
```diff
@@ -0,0 +1,1103 @@
...
+        self.ready.put(Some(true));
+    }
+
+
+    pub fn get_ready_status(&self) -> bool {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: more rusty name would be `is_ready`

File: src/chips/sam4l/flashcalw.rs [hunk 18]
```diff
@@ -0,0 +1,1103 @@
...
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
+    registers: *mut Registers,
+    ahb_clock: pm::Clock,
+    hramc1_clock: pm::Clock,
+    pb_clock: pm::Clock,
+    wait_until_ready: fn(&FLASHCALW) -> (),
+    error_status: TakeCell<u32>,
+    ready: TakeCell<bool>,
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: This should almost certainly be a `Cell` not a `TakeCell`, which would significantly simplify much of the code. In general, you should a `Cell` unless you _need_ a `TakeCell`. You typically only need a `TakeCell` when the internal value not `Copy` or is large.

File: src/chips/sam4l/flashcalw.rs [hunk 19]
```diff
@@ -0,0 +1,1103 @@
...
+            self.ready.put(Some(false));
+            false
+        } else {
+            self.ready.put(Some(true));
+            true
+        }
+    }
+    
+    const fn new(base_addr: *mut Registers, ahb_clk: pm::HSBClock,
+    hramc1_clk: pm::HSBClock, pb_clk: pm::PBBClock) -> FLASHCALW {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Indent

File: src/chips/sam4l/flashcalw.rs [hunk 20]
```diff
@@ -0,0 +1,1103 @@
...
+    Unconfigured
+}
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
+    registers: *mut Registers,
+    ahb_clock: pm::Clock,
+    hramc1_clock: pm::Clock,
+    pb_clock: pm::Clock,
+    wait_until_ready: fn(&FLASHCALW) -> (),
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: don't need this

File: src/chips/sam4l/flashcalw.rs [hunk 21]
```diff
@@ -0,0 +1,1103 @@
...
+    WritePageBuffer,
+    Writing,
+    Reading,
+    Erasing,
+    Ready,
+    Unconfigured
+}
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: In general, it's not clear what the use of several of the state fields are. I think some of them could/should be removed, but those you choose to retain, it would be useful to comment what they are used for and what invariant should be maintained.

File: src/chips/sam4l/flashcalw.rs [hunk 22]
```diff
@@ -0,0 +1,1103 @@
...
+}
+
+//static instance for the board. Only one FLASHCALW on chip.
+pub static mut flash_controller : FLASHCALW = 
+    FLASHCALW::new(FLASHCALW_BASE_ADDRS, pm::HSBClock::FLASHCALW, 
+        pm::HSBClock::FLASHCALWP, pm::PBBClock::FLASHCALW);
+
+
+// Few constants relating to module configuration.
+const FLASH_PAGE_SIZE : u32 = 512;
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: For all these constants, i think it's clearer without the `FLASH_` prefix. We're already in the flashcalw module, so we know what they are related too. `FLASH_` style prefixes are a relic from C, which doesn't have a namespace system.

File: src/chips/sam4l/flashcalw.rs (hunk 17 above)
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: This method is _probably_ not necessary, but if it is, it's _waaaay_ over complicated. Changing `self.ready` to a `Cell` fixes this I believe.

File: src/chips/sam4l/flashcalw.rs [hunk 23]
```diff
@@ -0,0 +1,1103 @@
...
+            client:                 TakeCell::empty(),
+            current_state:          Cell::new(FlashState::Unconfigured),
+            current_command:        Cell::new(Command::None),
+            page:                   Cell::new(0),
+            page_buffer:            TakeCell::new([0; FLASH_PAGE_SIZE as usize])
+        }
+    }
+
+    // Helper to read a flashcalw register 
+    fn read_register(&self, key : RegKey) -> u32 {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: code calling this should just directly dereference the registers themselves, it's much clearer and almost certainly way more efficient, unless the compiler does some serious in-lining (which it might)

File: src/chips/sam4l/flashcalw.rs [hunk 24]
```diff
@@ -0,0 +1,1103 @@
...
+    pub fn handle_interrupt(&self) {
+        unsafe { 
+            //  mark the controller as ready and clear pending interrupt 
+            self.mark_ready();  
+            nvic::clear_pending(nvic::NvicIdx::HFLASHC);
+        }
+       
+        self.error_status.put(Some(self.get_error_status()));
+        let error_status = self.error_status.take().unwrap();
+        self.error_status.put(Some(error_status));
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: are line 312+313 doing anything?

File: src/chips/sam4l/flashcalw.rs [hunk 25]
```diff
@@ -0,0 +1,1103 @@
...
+        self.issue_command(FlashCMD::QPR, page_number);
+        self.is_page_erased()
+    }
+
+    pub fn flashcalw_erase_page(&self, page_number : i32, check : bool) -> bool {
+        let mut page_erased = true;
+
+        self.issue_command(FlashCMD::EP, page_number);
+        if check {
+            let mut error_status : u32 = self.error_status.take().unwrap();
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Yeah, `self.error_status` definitely needs to be a `Cell`, you're emulating a `Cell` everywhere you use it.

File: src/chips/sam4l/flashcalw.rs [hunk 26]
```diff
@@ -0,0 +1,1103 @@
...
+            //  mark the controller as ready and clear pending interrupt 
+            self.mark_ready();  
+            nvic::clear_pending(nvic::NvicIdx::HFLASHC);
+        }
+       
+        self.error_status.put(Some(self.get_error_status()));
+        let error_status = self.error_status.take().unwrap();
+        self.error_status.put(Some(error_status));
+
+        //  Since the only interrupt request on is FRDY, a command should have
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: s/on is/is on/

File: src/chips/sam4l/flashcalw.rs [hunk 27]
```diff
@@ -0,0 +1,1103 @@
...
+        //  either completed or failed at this point.
+
+        // Check for errors and report to Client if there are
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: "... if there are __any__"

Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: or "... and report any to the Client"

File: src/chips/sam4l/flashcalw.rs [hunk 28]
```diff
@@ -0,0 +1,1103 @@
...
+        }
+        
+        //  Part of a command succeeded -- continue onto next steps.
+         
+        match self.current_command.get() {
+            Command::Write => {
+                match self.current_state.get() {
+                    FlashState::Unlocking => {
+                        self.current_state.set(FlashState::Erasing);
+                        self.flashcalw_erase_page(self.page.get(), true);
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: seems to me like, in general, you're using state variables like the page number in conjunction with particular in progress commands. It's equivalent, and IMO more semantically meaningful, to store these alongside either the `current_command` or `current_state`, e.g.:

```rust
enum Command {
    Write { page: u32 },
    Read { page: u32 },
    ...
}
```

File: src/chips/sam4l/flashcalw.rs [hunk 29]
```diff
@@ -0,0 +1,1103 @@
...
+                    },
+                    FlashState::Locking => {
+                        self.current_state.set(FlashState::Ready);
+                        self.current_command.set(Command::None);
+                    },
+                    _ => { assert!(false) /* should never reach here */ }
+
+                }
+            },
+            Command::Read => {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: yeah, just remove this variant from `Command`

File: src/chips/sam4l/flashcalw.rs [hunk 30]
```diff
@@ -0,0 +1,1103 @@
...
+    }
+
+    pub fn get_page_count_per_region(&self) -> u32 {
+        self.get_page_count() / FLASH_NB_OF_REGIONS
+    }
+
+
+    pub fn get_page_region(&self, page_number : i32) -> u32 {
+        (if page_number >= 0 
+            { unsafe { mem::transmute(page_number) } }
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: oh gosh, no! mem::transmute is super unsafe and should only be used when doing exceptional things like casting MMIO to a struct or something else not modeled in the language. This is just `page_number as u32`.

File: src/chips/sam4l/flashcalw.rs [hunk 31]
```diff
@@ -0,0 +1,1103 @@
...
+        / self.get_page_count_per_region())
+    }
+
+    pub fn get_region_first_page_number(&self, region : u32) -> u32 {
+        region * self.get_page_count_per_region()    
+    }
+
+
+    /// FLASHC Control
+    fn change_control_single_bit_val(&self, position : u32, enable : bool) {
```
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: I would just do thine simple one liner in each function that calls this instead. This function is a bit hard to parse

**Comment by alevy** [comment]
In general, the Flashcalw impl is very comprehensive, but on the other hand it's kind of hard to see the woods from the trees. A lot of exposed methods seems unnecessary. What is the interface I actually want to use as a client? For example, why would a Client every call `is_ready_int_enabled`? I would simplify it down to the necessary parts. We can always add simple methods like that if/when we need them.

File: src/chips/sam4l/flashcalw.rs (hunk 30 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: whoops 👍 . I remember I had read somewhere that you couldn't typecast between types of same size (i.e. u32 -> i32) but that was bs.

File: src/chips/sam4l/flashcalw.rs (hunk 24 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: Just reorganized.
```
let error_status= self.get_error_status();
self.error_status.put(Some(error_status));
```
The local error_status is used a few lines below to tell the client if there was an error, what error it was.

File: src/chips/sam4l/flashcalw.rs (hunk 26 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: 'Since the only interrupt on is FRDY' 🎯

File: src/chips/sam4l/flashcalw.rs (hunk 13 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: Was originally in the Atmel implementation, since by default they did polling on a particular bit. Since our implementation uses interrupts we don't need it and I will remove it. It was useful though, as if it was ever called then the flash wasn't being used correctly.

File: src/chips/sam4l/flashcalw.rs (hunk 26 above)
Comment (alevy) [inline src/chips/sam4l/flashcalw.rs]: Oh... Duh. Nm

File: src/chips/sam4l/flashcalw.rs (hunk 17 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: yep that function was a mess. Deleted it.

File: src/chips/sam4l/flashcalw.rs (hunk 23 above)
Comment (KBaichoo) [inline src/chips/sam4l/flashcalw.rs]: talked with @alevy about this. With optimizations, the compiler would know which branch we'll take, and with inlining setting up the stack frame isn't much work. Using this method in places where there's only one read from the registers is about as good as having it everywhere + it reuses code. So think I'll keep this one.
